### DISCUSSION TITLE
supports sparse vector of a quantum state to evaluate expectation value

### DISCUSSION THREAD

**Discussion by ywlds** [QuantumBFS/Yao.jl#296] [Q&A]
for some case we want to evaluate <\psi| H |\psi>.
it seems the mat(H) can get the sparse matrix, while the statevec(\psi) is a dense vector.

How can we make them both sparse to accelerate the numerical calculations?

Many thanks.

**Comment by Roger-luo** [comment]
If you want to apply gates on the state psi, using a sparse vector won't help much - it will eventually grow to a dense vector anyway.

if you just want to calculate the expectation of a known sparse state, you should directly construct the known state in `SparseVector` and wrap it in the `ArrayReg` - the `ArrayReg` accepts any kind of Julia arrays.
